Compiler bug when try to add a structure in a enum field

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Manjaro Linux"
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CC version: cc (GCC) 11.1.0

getwd: /var/www/vproject
vmodules: /home/tanguy/.vmodules
vroot: /var/www/vproject/vorigin
vexe: /var/www/vproject/vorigin/v
vexe mtime: 2021-08-23 07:20:48
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 51d7aed

Git version: git version 2.32.0
Git vroot status: weekly.2021.33.2-58-g51d7aede (4 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg /var/www/vproject/scarp/scrap_bug.v`
```v
module main

struct Node {
mut:
    tag string
    content []Content
}

struct Text {
mut:
    content string
}

type Content = Text | Node

fn main() {
  
    mut body := Node{tag: "body", content:[]Content{}}
    children := Node{tag: "p", content:[]Content{}}

    body.content.insert(0, children)
    println(body)
}

```

**What did you expect to see?**

I don't know if I should see an error or not.


**What did you see instead?**
```
==================
/tmp/v/scrap_bug.4195359145150962878.tmp.c:10433: error: cannot convert 'struct main__Node' to 'struct main__Text *'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -411,7 +411,7 @@ fn (mut g Gen) gen_array_insert(node ast.CallExpr) {
 		if left_info.elem_type == ast.string_type {
 			g.write('string_clone(')
 		}
-		g.expr(node.args[1].expr)
+		g.expr_with_cast(node.args[1].expr, node.args[1].typ, left_info.elem_type)
 		if left_info.elem_type == ast.string_type {
 			g.write(')')
 		}
@@ -441,7 +441,7 @@ fn (mut g Gen) gen_array_prepend(node ast.CallExpr) {
 		g.write('.len)')
 	} else {
 		g.write(', &($elem_type_str[]){')
-		g.expr(node.args[0].expr)
+		g.expr_with_cast(node.args[0].expr, node.args[0].typ, left_info.elem_type)
 		g.write('})')
 	}
 }

--- a/vlib/v/tests/array_of_sumtypes_test.v
+++ b/vlib/v/tests/array_of_sumtypes_test.v
@@ -1,0 +1,20 @@
+struct Node {
+mut:
+	tag     string
+	content []Content
+}
+
+struct Text {
+}
+
+type Content = Node | Text
+
+fn test_push_prepend_insert() {
+	mut body := Node{}
+	body.content << Node{
+		tag: 'a'
+	}
+	body.content.prepend(Node{ tag: 'b' })
+	body.content.insert(1, Node{ tag: 'c' })
+	assert body.content.len == 3
+}

